src/snowflake/snowpark/dataframe.py

```diff
@@ -581,8 +580,13 @@ def set_ast_ref(self, sp_dataframe_expr_builder: Any) -> None:
         """
         Given a field builder expression of the AST type SpDataframeExpr, points the builder to reference this dataframe.
         """
+        if self._ast_id is None and FAIL_ON_MISSING_AST:
+            _logger.debug(self._explain_string())
+            raise NotImplementedError(
+                f"DataFrame with API usage {self._plan.api_calls} is missing complete AST logging."
+            )
         # TODO: remove the None guard below once we generate the correct AST.
-        if self._ast_id is not None:
+        elif self._ast_id is not None:
```

Comment:
This should be just an `else`. You should fold the FAIL_ON_MISSING_AST_CHECK above into a nested if in the `then` branch of the `if` statement.